### PR TITLE
Reconstruct HTTP Result in safe way using Result.copy().

### LINKED ISSUE #64: 0.7.0 silently kills the session, flash and cookies
Look at [line 92 in CompressorFilter.scala](https://github.com/mohiva/play-html-compressor/blame/master/app/com/mohiva/play/compressor/CompressorFilter.scala#L92):
```
 case HttpEntity.Strict(data, contentType) =>
    Future.successful(Result(result.header, HttpEntity.Strict(ByteString(compress(data)), contentType)))
```
`Result()` constructor called with 2 args, but constructor signature have **5** args (3 optional):
```
case class Result(header: ResponseHeader, body: HttpEntity,
    newSession: Option[Session] = None, newFlash: Option[Flash] = None, newCookies: Seq[Cookie] = Seq.empty) {
```
`HtmlCompressor` filter reconstructs the results using headers and body, but without passing other arguments, like `newSession` and others.

Maybe to use `Result.copy(header=..., body=...)`?
It will be safer, and protect from so hard errors in future.

### PR DESCRIPTION
**Pull request by helllamer**
This prevents losing some parts of Result like session and other.
Fixes #64

### CODE REVIEW

File: app/com/mohiva/play/compressor/CompressorFilter.scala [hunk 1]
```diff
@@ -88,17 +88,31 @@ abstract class CompressorFilter[C <: Compressor] extends Filter {
...
+            bytes <- body0.data
+              .toMat(Sink.fold(ByteString())(_ ++ _))(Keep.right).run()
+          } yield {
             val compressed = compress(bytes)
             val length = compressed.length
-            Result(
-              result.header.copy(headers = result.header.headers),
-              HttpEntity.Streamed(Source.single(ByteString(compressed)), Some(length.toLong), result.body.contentType)
+            result.copy(
+              //header = result.header.copy(headers = result.header.headers),
```
Comment (jleider) [inline app/com/mohiva/play/compressor/CompressorFilter.scala]: Did you mean to leave this commented out?

Comment (helllamer) [inline app/com/mohiva/play/compressor/CompressorFilter.scala]: Hmm, this code previously was UNcommented. But i think, this code makes no sense.
May be, something was or invalid missing here?

**Comment by helllamer** [comment]
Ok, i've removed couple of useless lines from project. Not so much to review.

File: app/com/mohiva/play/compressor/CompressorFilter.scala [hunk 2]
```diff
@@ -88,17 +88,30 @@ abstract class CompressorFilter[C <: Compressor] extends Filter {
 
     if (isCompressible(result)) {
       result.body match {
-        case HttpEntity.Strict(data, contentType) =>
-          Future.successful(Result(result.header, HttpEntity.Strict(ByteString(compress(data)), contentType)))
-        case HttpEntity.Streamed(data, _, contentType) =>
-          data.toMat(Sink.fold(ByteString())(_ ++ _))(Keep.right).run() map { bytes =>
+
```
Comment (akkie) [inline app/com/mohiva/play/compressor/CompressorFilter.scala]: I've stumbled over the same issue in Silhouette. Could you please change the code to something like this:
```scala
      result.body match {
        case body: HttpEntity.Strict =>
          Future.successful(result.copy(
            body = body.copy(
              data = ByteString(compress(body.data))
            )
          ))
        case body: HttpEntity.Streamed =>
          for {
            bytes <- body.data.toMat(Sink.fold(ByteString())(_ ++ _))(Keep.right).run()
          } yield {
            val compressed = compress(bytes)
            val length = compressed.length
            result.copy(
              body = body.copy(
                data = Source.single(ByteString(compressed)),
                contentLength = Some(length.toLong)
              )
            )
          }
        case _ => 
          Future.successful(result)
      }
```

Appending number to variables is a code smell in my eyes :wink:

**Comment by helllamer** [comment]
Ok, the *last* fix.

File: app/com/mohiva/play/compressor/CompressorFilter.scala [hunk 3]
```diff
@@ -88,15 +88,25 @@ abstract class CompressorFilter[C <: Compressor] extends Filter {
 
     if (isCompressible(result)) {
       result.body match {
-        case HttpEntity.Strict(data, contentType) =>
-          Future.successful(Result(result.header, HttpEntity.Strict(ByteString(compress(data)), contentType)))
-        case HttpEntity.Streamed(data, _, contentType) =>
-          data.toMat(Sink.fold(ByteString())(_ ++ _))(Keep.right).run() map { bytes =>
+        case body0: HttpEntity.Strict =>
```
Comment (akkie) [inline app/com/mohiva/play/compressor/CompressorFilter.scala]: Why do you use `body0` instead of `body`?

**Comment by helllamer** [comment]
Why you are spending minutes your (and my) time for this endless comments and bla-bla instead of just-fix-it-as-YOU-want in several seconds?

**Comment by akkie** [comment]
First of all thanks for your contribution. The reason is that I try to keep the code quality high. I've explained it yesterday, that in my opinion,  attaching numbers to variable names is a code smell. So I don't understand why you attach the numbers again?

**Comment by helllamer** [comment]
1. Frankly speaking, I don't see a "code quality" in the sources. Maybe, we a read different sources?
Some code lines looking like "minified js", where mostly all newlines and spaces are stripped out with fences of braces.
During years, this silently causes crazy errors like #65 and deadly useless code like `result.header.copy(headers = result.header.headers)`, because very-long-horizontal-lines-what-do-everything are hard to read by 3rd-party people, most people prefer not to read such code, only to patch it.

2. Second problem, i already implemented what i needed, and modified the code by your, guys, non-constructive wishes. But, let's stop here. After second code-style-wish, it was clear to me: code is already ok, problem not in the code.

**Comment by akkie** [comment]
Thanks for your contribution. I'll change the lines by myself.

**Pull request merged**